src/control/cmd/daos/util.c

```diff
@@ -42,56 +42,63 @@ resolve_duns_path(struct cmd_args_s *ap)
 	char   *name = NULL, *dir_name = NULL;
 
 	rc = duns_resolve_path(ap->path, &dattr);
-	if (rc == 0) {
-		ap->type = dattr.da_type;
-		uuid_copy(ap->p_uuid, dattr.da_puuid);
-		uuid_copy(ap->c_uuid, dattr.da_cuuid);
-	} else {
-		if (ap->fs_op == -1) {
-			rc = call_dfuse_ioctl(ap->path, &il_reply);
-			if (rc == 0) {
-				ap->type = DAOS_PROP_CO_LAYOUT_POSIX;
-				uuid_copy(ap->p_uuid, il_reply.fir_pool);
-				uuid_copy(ap->c_uuid, il_reply.fir_cont);
-				ap->oid = il_reply.fir_oid;
-			}
-		} else if (rc == ENOENT && ap->fs_op == FS_SET_ATTR) {
-			/** we could be creating a new file, so try dirname */
-			parse_filename_dfs(ap->path, &name,
-				&dir_name);
-
-			rc = duns_resolve_path(dir_name, &dattr);
-			if (rc == 0) {
-				ap->type = dattr.da_type;
-				uuid_copy(ap->p_uuid, dattr.da_puuid);
-				uuid_copy(ap->c_uuid, dattr.da_cuuid);
-			}
-		}
+	if (rc == ENOENT && ap->fs_op == FS_SET_ATTR) {
+		/** we could be creating a new file, so try dirname */
+		parse_filename_dfs(ap->path, &name, &dir_name);
+
+		rc = duns_resolve_path(dir_name, &dattr);
 	}
 
-	if (rc != 0) {
-		fprintf(ap->errstream, "could not resolve "
-		"pool, container by "
-		"path: %d %s %s\n",
-		rc, strerror(rc), ap->path);
+	if (rc && ap->fs_op == -1) {
+		rc = call_dfuse_ioctl(ap->path, &il_reply);
+		if (rc == 0) {
+			ap->type = DAOS_PROP_CO_LAYOUT_POSIX;
+			uuid_copy(ap->p_uuid, il_reply.fir_pool);
+			uuid_copy(ap->c_uuid, il_reply.fir_cont);
+			ap->oid = il_reply.fir_oid;
+			D_GOTO(out, rc);
+		}
+	}
 
+	if (rc) {
+		fprintf(ap->errstream, "could not resolve pool, container by "
+		"path: %d %s %s\n", rc, strerror(rc), ap->path);
 		D_GOTO(out, rc);
 	}
 
+	ap->type = dattr.da_type;
+
+	/** set pool/cont label or uuid */
+	if (dattr.da_pool_label) {
+		D_STRNDUP(ap->pool_label, dattr.da_pool_label,
+			  DAOS_PROP_LABEL_MAX_LEN);
+		if (ap->pool_label == NULL)
+			D_GOTO(out, rc);
+	} else {
+		uuid_copy(ap->p_uuid, dattr.da_puuid);
+	}
+
+	if (dattr.da_cont_label) {
+		D_STRNDUP(ap->cont_label, dattr.da_cont_label,
+			  DAOS_PROP_LABEL_MAX_LEN);
+		if (ap->cont_label == NULL)
+			D_GOTO(out, rc);
+	} else {
+		uuid_copy(ap->c_uuid, dattr.da_cuuid);
+	}
+
 	if (ap->fs_op != -1) {
 		if (name) {
 			if (dattr.da_rel_path) {
 				D_ASPRINTF(ap->dfs_path, "%s/%s",
-						dattr.da_rel_path, name);
-				free(dattr.da_rel_path);
+					   dattr.da_rel_path, name);
 			} else {
 				D_ASPRINTF(ap->dfs_path, "/%s", name);
 			}
 		} else {
 			if (dattr.da_rel_path) {
 				D_STRNDUP(ap->dfs_path,
-						dattr.da_rel_path, PATH_MAX);
-				free(dattr.da_rel_path);
+					  dattr.da_rel_path, PATH_MAX);
```

Comment:
In theory this could "steal" the string by copying the pointer and setting the old one to NULL.